packages/synapse-sdk/src/storage/context.ts

```diff
@@ -392,47 +392,64 @@ export class StorageContext {
     signerAddress: string,
     options: StorageServiceOptions
   ): Promise<ProviderSelectionResult> {
-    // Fetch data sets to find the specific one
-    const dataSets = await warmStorageService.getClientDataSetsWithDetails(signerAddress)
-    const dataSet = dataSets.find((ds) => ds.pdpVerifierDataSetId === dataSetId)
+    // Optimized approach: fetch only the single dataset we need instead of all client datasets
+    // This avoids expensive getClientDataSetsWithDetails() which does multicalls for ALL datasets
+
+    // Step 1: Validate that dataset is live and managed (throws if not)
+    await warmStorageService.validateDataSet(dataSetId)
+
+    // Step 2: Get basic dataset info and metadata in parallel
+    const [dataSetInfo, dataSetMetadata] = await Promise.all([
```

Comment:
you could even put `validateDataSet` inside the `Promise.all`, no need to block on it since you're just looking for a throw and then you get to parallelise it even more